datanommer.models/datanommer/models/__init__.py

```diff
@@ -128,7 +130,16 @@ def add(envelope):
     obj.msg = message['msg']
     obj.headers = headers
 
-    session.add(obj)
+    # Construct an INSERT ... SELECT ... which will do nothing if an
+    # entry with a matching msg_id already exists. This can happen in the
+    # event of message redelivery.
+    cols = [c for c in Message.__table__.c if c.name != 'id']
+    vals = [literal(getattr(obj, c.name)) for c in cols]
+    not_exists = ~exists(select([Message.__table__.c.i]).\
+                         where(Message.msg_id == msg_id))
+    insert = Message.__table__.insert().from_select(cols, select(vals).\
+                                                    where(not_exists))
+    session.execute(insert)
```

Comment:
I can understand the objection to the duplicate message handling, since that is an exceptional condition. I could rewrite that to do a try/catch. However, I do feel like the approach in get_or_create() is correct, and is the optimal way to handle concurrency when multiple workers are populating the same tables with singleton data.